src/dmd/sideeffect.d

```diff
@@ -332,16 +332,10 @@ bool discardValue(Expression e)
     case TOK.comma:
         {
             CommaExp ce = cast(CommaExp)e;
-            /* Check for compiler-generated code of the form  auto __tmp, e, __tmp;
-             * In such cases, only check e for side effect (it's OK for __tmp to have
-             * no side effect).
-             * See https://issues.dlang.org/show_bug.cgi?id=4231 for discussion
-             */
-            auto fc = firstComma(ce);
-            if (fc.op == TOK.declaration && ce.e2.op == TOK.variable && (cast(DeclarationExp)fc).declaration == (cast(VarExp)ce.e2).var)
-            {
+            // Don't complain about compiler-generated comma expressions
+            if (ce.isGenerated)
```

Comment:
Maybe add a sanity check that at least one of the expressions has a side effect, e.g. as

`assert(hasSideEffect(ce.e1) || hasSideEffect(ce.e2))`

Dmd shouldn't create `CommaExp` without any side effect.

Reply:
Good idea.

Reply:
Added the proposed `assert`.

Reply:
The assert has revealed that DMD creates a chain of commas for struct/class `invariant`, as these are functions with no side-effects the condition is not met.
I'm back to returning a plain `false` since the old deleted condition was just a pattern to detect the most common generated comma from a time where commas were allowed everywhere.